### PR TITLE
2827 Implemented dev cl-postgres SSL connection

### PR DESCRIPTION
**Pull request by albertisfu**
This PR adds support for SSL connections to cl-postgres for development.

- In order to add an SSL certificate to `cl-postgres`, it was necessary to create a Dockerfile. So that self-signed certificate is generated when the `cl-postgres` container is created, and is then activated in the `postgresql.conf` settings.
- `DB_SSL_MODE` is now set to require by default, thus SSL is used in development.
- Mounting the directory in `/root/.postgresql` isn't essential for operation, but I included it to emulate an error similar to the one that occurred in production. Since Docker containers run as root in development no permission error is raised, mounting an empty `postgresql.crt` file triggers an error due to the file being blank. So we can confirm that with the addition of the `PGSSLCERT=/tmp/postgresql.crt` environment variable in `cl-django` solves this problem as described in issue #2827.

- In summary, this change emulates the issue that occurred in production. By also using an SSL connection to `cl-postgres` in development, this has made the development environment more closely mirror the production database connection, so it'll help us catch any SSL-related errors in the future in development.

- **Before deploying this to prod it's required to set the `PGSSLCERT=/tmp/postgresql.crt` env var in production, assuming the default user in Kubernetes has read permissions to `/tmp/`**

### CODE REVIEW

**Comment by mlissner** [comment]
Hm, is there a way to do a smaller version of this without the self-signed cert and without all the extra scripts? I appreciate the effort, but it feels like a lot to understand when you just want to run docker in dev. For example, we don't have self-signed certs in prod.

**Comment by albertisfu** [comment]
Sure, it's a good idea to set the PGSSLCERT in the Dockerfile as you mentioned in #2827. I'll add it.

I think an alternative for generating a self-signed certificate on the fly, while still maintaining encryption in development, is to generate a self-signed certificate for `cl-postgres` once, then upload the cert files to the repository and mount them to the Postgres container. However, this method would still necessitate a custom Docker image for PostgreSQL in order to set the cert files. However, it will be simpler than generating the certs on the fly.

Let me know what you think about this approach.

Another alternative is to use a remote DB instance that already has encryption support, but I think that's not a viable solution since it'll require a DB for each dev and doesn't seem as performant as having a local DB.

**Comment by mlissner** [comment]
Sure, that seems like a better approach. Let's get this prioritized though, because we can't do deploys to CL until it's fixed.

**Comment by albertisfu** [comment]
I've applied the requested changes.

Now it utilizes a pre-generated certificate with an expiration of 825 days; it will need to be regenerated and replaced after this time:
`openssl req -newkey rsa:2048 -nodes -x509 -keyout cl-postgres.key -out cl-postgres.crt -days 825 -subj /CN=cl-postgres`

- PGSSLCERT=/tmp/postgresql.crt environment variable is now set in the Dockerfile.
- Added an exception to `detect-private-key` pre-commit hook to allow `cl-postgres.req` key in the repo.

**Comment by mlissner** [comment]
Why 825 days instead of 100,000?

**Comment by albertisfu** [comment]
Yeah, the 825-day expiration was a reference to a duration that most certificate providers used in the past. Checking in detail, this expiration time has been [reduced](https://www.ssl.com/blogs/ssl-certificate-maximum-duration-825-days/) to 398 days. However, this lifetime appears to only impact major browsers; it doesn't seem as though the 'postgrestsql' client adheres to this policy.

In theory, we can generate a certificate with an expiration time as long as it can be encoded in the certificate. I've already updated the certificate with an expiration period of 100,000 days. In practice, I believe this certificate will need to be renewed before the expiration time, due to protocol updates.

**Comment by mlissner** [comment]
>  In practice, I believe this certificate will need to be renewed before the expiration time, due to protocol updates.

Perhaps...

image

**Comment by mlissner** [comment]
All set. We're back.
